library/init/meta/tactic.lean

```diff
@@ -1411,7 +1413,84 @@ meta def add_defn_equations (lp : list name) (params : list expr) (fn : expr)
 do opt ← get_options,
    updateex_env $ λ e, e.add_defn_eqns opt lp params fn eqns is_meta
 
+/-- Get the revertible part of the local context. These are the hypotheses that
+appear after the last frozen local instance in the local context. We call them
+revertible because `revert` can revert them, unlike those hypotheses which occur
+before a frozen instance. -/
+meta def revertible_local_context : tactic (list expr) :=
+do ctx ← local_context,
+   frozen ← frozen_local_instances,
+   pure $
```

Comment:
Is this different from `ctx.drop (frozen_local_instance.get_or_else []).length`?

Reply:
I think so, yes. Your version drops as many hyps from the context as there are frozen instances. However, we can have other hyps in between the frozen instances, so the part of the context that we can't touch can be longer.

Reply:
That's my question: is this possible?

Reply:
Yes, here's a full test case where the implementations diverge.

```lean 
namespace list

def after {α} (p : α → Prop) [decidable_pred p] : list α → list α
| [] := []
| (x :: xs) := if p x then xs else after xs

end list

namespace tactic

meta def revertible_local_context₁ : tactic (list expr) :=
do ctx ← local_context,
   frozen ← frozen_local_instances,
   pure $
     match frozen with
     | none := ctx
     | some [] := ctx
     | some (h :: _) := ctx.after (eq h)
     end

meta def revertible_local_context₂ : tactic (list expr) :=
do ctx ← local_context,
   frozen ← frozen_local_instances,
   pure $ ctx.drop $ (frozen.get_or_else []).length

meta def interactive.test : tactic unit :=
do revertible_local_context₁ >>= trace,
   revertible_local_context₂ >>= trace

end tactic


lemma test {α : Prop} [decidable α] {β : Prop} [decidable β] : unit :=
begin
  test,
  exact ()
end
```

Reply:
Ah, so `frozen_local_instances` only returns the instances.